kubernetes/config/kube_config_test.py

```diff
@@ -0,0 +1,413 @@
+        self.assertEqual("test_obj_with_names[name=test_name2]",
+                         node.get_with_name("test_name2").name)
+        self.assertEqual("test_obj_with_names[name=test_name3]",
+                         node.get_with_name("test_name3").name)
+
+    def expect_exception(self, func, message_part):
+        try:
+            func()
+            self.fail("Should fail.")
+        except ConfigException as e:
+            self.assertTrue(
+                message_part in e.message, "'%s' should be in '%s'" %
+                (message_part, e.message))
+
+    def test_invalid_key(self):
+        node = ConfigNode("test_obj", self.test_obj)
+        self.expect_exception(lambda: node['non-existance-key'],
+                              "Expect key non-existance-key in test_obj")
+        self.expect_exception(lambda: node['key3']['non-existance-key'],
+                              "Expect key non-existance-key in test_obj/key3")
+        self.expect_exception(
+            lambda: node['key2'].get_with_name('noname'),
+            "Expect all values in test_obj/key2 list to have \'name\' key")
+        self.expect_exception(
+            lambda: node['key3'].get_with_name('noname'),
+            "Expect test_obj/key3 to be a list")
+        self.expect_exception(
+            lambda: node['with_names'].get_with_name('noname'),
+            "Expect object with name noname in test_obj/with_names list")
+
+
+class FakeConfig:
+
+    FILE_KEYS = ["ssl_ca_cert", "key_file", "cert_file"]
+
+    def __init__(self, token=None, **kwargs):
+        self.api_key = {}
+        if token:
+            self.api_key['authorization'] = token
+
+        self.__dict__.update(kwargs)
+
+    def __eq__(self, other):
+        if len(self.__dict__) != len(other.__dict__):
+            return
+        for k, v in self.__dict__.items():
+            if k not in other.__dict__:
+                return
+            if k in self.FILE_KEYS:
+                try:
+                    with open(v) as f1, open(other.__dict__[k]) as f2:
+                        if f1.read() != f2.read():
+                            return
+                except IOError as e:
+                    # fall back to only compare filenames in case we are testing
+                    # passing filename to the config
+                    if other.__dict__[k] != v:
+                        return
+            else:
+                if other.__dict__[k] != v:
+                    return
+        return True
+
+    def __repr__(self):
+        rep = "\n"
+        for k, v in self.__dict__.items():
+            val = v
+            if k in self.FILE_KEYS:
+                try:
+                    with open(v) as f:
+                        val = "FILE: %s" % str.decode(f.read())
+                except IOError as e:
+                    val = "ERROR: %s" % str(e)
+            rep += "\t%s: %s\n" % (k, val)
+        return "Config(%s\n)" % rep
+
+
+class TestKubeConfigLoader(TestRoot):
```

Comment:
I will add more unit tests for lower level functions, but keep these ones as it test the overall behaviour.